Prepare DisplayLink to support UI-side compositing

smfr opened the pull request:
#### 0ca300a709eed3e70130a5e13cab0e5b5887b9b6
<pre>
Prepare DisplayLink to support UI-side compositing
<a href="https://bugs.webkit.org/show_bug.cgi?id=246694">https://bugs.webkit.org/show_bug.cgi?id=246694</a>
rdar://101296426

Reviewed by Chris Dumez.

The DisplayLink class has clients (currently denoted by IPC::Connection::UniqueID), and each client
has a collection of DisplayLinkObserverID, whose values are only unique in that context of that client.
In addition, each client has some extra state in the form of &quot;wants full speed updates&quot; which is used
for 120Hz scrolling animations and event handling.

With UI-side compositing, we&apos;ll need a client in the UI process. So replace IPC::Connection::UniqueID
with a Client class. For web process clients, DisplayLinkProcessProxyClient implements DisplayLink::Client,
and handles the IPC to the web process. WebProcessProxy owns an instance of DisplayLinkProcessProxyClient.
Getting WebProcessProxy&apos;s IPC connection isn&apos;t thread-safe, so we cache the IPC::Connection on
the main thread. DisplayLink::Client is accessed via a thread-safe CheckedPtr&lt;&gt; to catch lifetime
issues.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::startDisplayLink):
(WebKit::WebProcessPool::stopDisplayLink):
(WebKit::WebProcessPool::stopDisplayLinks):
(WebKit::WebProcessPool::setDisplayLinkPreferredFramesPerSecond):
(WebKit::WebProcessPool::setDisplayLinkForDisplayWantsFullSpeedUpdates):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateDisplayLinkFrequency):
(WebKit::WebPageProxy::windowScreenDidChange):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::~WebProcessProxy):
(WebKit::WebProcessProxy::connectionWillOpen):
(WebKit::WebProcessProxy::processWillShutDown):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::displayLinkClient):
* Source/WebKit/UIProcess/mac/DisplayLink.cpp:
(WebKit::DisplayLink::addObserver):
(WebKit::DisplayLink::removeObserver):
(WebKit::DisplayLink::removeClient):
(WebKit::DisplayLink::removeInfoForClientIfUnused):
(WebKit::DisplayLink::incrementFullSpeedRequestClientCount):
(WebKit::DisplayLink::decrementFullSpeedRequestClientCount):
(WebKit::DisplayLink::setObserverPreferredFramesPerSecond):
(WebKit::DisplayLink::notifyObserversDisplayWasRefreshed):
(WebKit::DisplayLink::removeObservers): Deleted.
(WebKit::DisplayLink::removeInfoForConnectionIfPossible): Deleted.
(WebKit::DisplayLink::setPreferredFramesPerSecond): Deleted.
* Source/WebKit/UIProcess/mac/DisplayLink.h:
(WebKit::DisplayLink::displayID const):
(WebKit::DisplayLink::setShouldSendIPCOnBackgroundQueue): Deleted.
* Source/WebKit/UIProcess/mac/DisplayLinkProcessProxyClient.cpp: Added.
(WebKit::DisplayLinkProcessProxyClient::setConnection):
(WebKit::DisplayLinkProcessProxyClient::displayLinkFired):
* Source/WebKit/UIProcess/mac/DisplayLinkProcessProxyClient.h: Added.
* Source/WebKit/UIProcess/mac/WebProcessProxyMac.mm:
(WebKit::WebProcessProxy::startDisplayLink):
(WebKit::WebProcessProxy::stopDisplayLink):
(WebKit::WebProcessProxy::setDisplayLinkPreferredFramesPerSecond):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDisplayRefreshMonitor.mm:
(WebKit::RemoteLayerTreeDisplayRefreshMonitor::requestRefreshCallback):

Canonical link: <a href="https://commits.webkit.org/255970@main">https://commits.webkit.org/255970@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dbebc23921182b608d447844b451fc0a8027fe56

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94192 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3384 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/24719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/103873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3403 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31598 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/86551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/99849 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99847 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2445 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/80606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/86551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/84369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/24719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/86551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37980 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/24719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35865 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/24719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4132 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39740 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/80606 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41682 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/24719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/81cd90db246e75251ce10607640e72b6eced894f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/5496

shallawa commented on Source/WebKit/UIProcess/mac/DisplayLink.h:
```diff
@@ -27,70 +27,71 @@
     Lock m_observersLock;
-    HashMap<IPC::Connection::UniqueID, ConnectionClientInfo> m_observers WTF_GUARDED_BY_LOCK(m_observersLock);
-    WebCore::PlatformDisplayID m_displayID;
+    HashMap<Client*, ClientInfo> m_observers WTF_GUARDED_BY_LOCK(m_observersLock);
```
Why is called `m_observers` although it maps from a `Client` to a `ClientInfo`? Should not it be called `m_clients`?

Why do we use a raw pointer for the key? Should we use WeakPtr<> instead? How do we guarantee the clients will remove themselves before they are destroyed.

smfr replied:
Done.

shallawa commented on Source/WebKit/UIProcess/mac/DisplayLink.cpp:
```diff
@@ -86,16 +84,16 @@ WebCore::FramesPerSecond DisplayLink::nominalFramesPerSecondFromDisplayLink(CVDi
         Locker locker { m_observersLock };
-        m_observers.ensure(connectionID, [] {
-            return ConnectionClientInfo { };
+        m_observers.ensure(&client, [] {
+            return ClientInfo { };
         }).iterator->value.observers.append({ observerID, preferredFramesPerSecond });
```
I think it will make changes a lot clearer if we rename `m_observers` to `m_clients`.

smfr replied:
Done.

shallawa commented on Source/WebKit/UIProcess/mac/DisplayLink.cpp:
```diff
@@ -110,97 +108,97 @@ void DisplayLink::addObserver(IPC::Connection::UniqueID connectionID, DisplayLin
 }
 
-void DisplayLink::removeInfoForConnectionIfPossible(IPC::Connection::UniqueID connectionID)
+void DisplayLink::removeInfoForClientIfPossible(Client& client)
```
Maybe we can just use `removeInfoForClient()` and we can return bool to indicate whether the info was removed or not. This is what `HashMap::remove()`.

smfr replied:
I can do this.

shallawa commented on Source/WebKit/UIProcess/mac/DisplayLink.h:
```diff
@@ -27,70 +27,71 @@
 #include <WebCore/DisplayUpdate.h>
 #include <WebCore/PlatformScreen.h>
-#include <wtf/HashMap.h>
+#include <wtf/HashSet.h>
```
Why this change is necessary? This class is only using HashMap.

smfr replied:
Leftover from a previous version of the patch.

shallawa commented on Source/WebKit/UIProcess/mac/DisplayLink.h:
```diff
@@ -27,70 +27,71 @@
-    // When responsiveness is critical, we send the IPC to a background queue. Otherwise, we send it to the
-    // main thread to avoid unnecessary thread hopping and save power.
-    static void setShouldSendIPCOnBackgroundQueue(bool value) { shouldSendIPCOnBackgroundQueue = value; }
+    void addClient(Client&);
+    void removeClient(Client&);
```
Where are these functions implemented? I do not see they are ever called.

smfr replied:
`removeClient()` is called on web process teardown. `addclient()` was never implemented; removed.

shallawa commented on Source/WebKit/Shared/DisplayLinkObserverID.h:
```diff
@@ -32,4 +32,7 @@ namespace WebKit {
 enum DisplayLinkObserverIDType { };
 using DisplayLinkObserverID = ObjectIdentifier<DisplayLinkObserverIDType>;
 
+enum DisplayLinkObserverCollectionIDType { };
+using DisplayLinkObserverCollectionID = ObjectIdentifier<DisplayLinkObserverCollectionIDType>;
```
Where is enum used? I do not see it in this PR.

smfr replied:
Agree, this is leftover.

shallawa commented on Source/WebKit/UIProcess/mac/DisplayLinkProcessProxyClient.h:
```diff
@@ -0,0 +1,57 @@
+    void displayLinkFired(WebCore::PlatformDisplayID, WebCore::DisplayUpdate, bool wantsFullSpeedUpdates, bool anyObserverWantsCallback) override;
+
+    Lock m_connectionIDLock;
+    IPC::Connection::UniqueID m_connectionID WTF_GUARDED_BY_LOCK(m_connectionIDLock);
```
Can't std::atomic<IPC::Connection::UniqueID> be used instead? It looks from the source file, we care only about the atomicity of setting and getting the value of this member.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4b2c8a4c8bd5f412fadd57107333c2cc2691e0a0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/5496

smfr commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -4084,7 +4084,7 @@ void WebPageProxy::windowScreenDidChange(PlatformDisplayID displayID, std::optio
 #if HAVE(CVDISPLAYLINK)
     if (hasRunningProcess() && m_displayID && m_registeredForFullSpeedUpdates)
-        process().processPool().setDisplayLinkForDisplayWantsFullSpeedUpdates(*m_process->connection(), *m_displayID, false);
+        process().processPool().setDisplayLinkForDisplayWantsFullSpeedUpdates(m_process.get(), *m_displayID, false);
```
Done.

kkinnunen-apple commented on Source/WebKit/UIProcess/mac/DisplayLink.cpp:
```diff
@@ -223,22 +224,20 @@ void DisplayLink::notifyObserversDisplayWasRefreshed()
     bool anyConnectionHadObservers = false;
-    for (auto& [connectionID, connectionInfo] : m_observers) {
-        if (connectionInfo.observers.isEmpty())
+    for (auto [client, clientInfo] : m_clients) {
```
[1] here you deref the weakptr in a wrong thread.

kkinnunen-apple commented on Source/WebKit/UIProcess/mac/DisplayLink.h:
```diff
@@ -27,70 +27,71 @@
-    Lock m_observersLock;
-    HashMap<IPC::Connection::UniqueID, ConnectionClientInfo> m_observers WTF_GUARDED_BY_LOCK(m_observersLock);
-    WebCore::PlatformDisplayID m_displayID;
+    Lock m_clientsLock;
```
I know it's an existing issue, but:
It is mostly a logical oxymoron to have a weak ptr and a lock together.
Weak ptrs are not thread-safe, and can only be dereferenced on the owner thread.
Then code down the line [1] dereferences the weak ptr in a wrong thread.

smfr replied:
Replaced the WeakPtr with CheckedPtr.

kkinnunen-apple commented on Source/WebKit/UIProcess/mac/DisplayLinkProcessProxyClient.cpp:
```diff
@@ -0,0 +1,58 @@
+        return;
+
+    if (wantsFullSpeedUpdates)
+        IPC::Connection::send(connectionID, Messages::EventDispatcher::DisplayWasRefreshed(displayID, displayUpdate, anyObserverWantsCallback), 0, { }, Thread::QOS::UserInteractive);
```
I know it's an existing problem, but:
send via `UniqueID` is not thread-safe. It will potentially revive a 0 refcount object and cause UAF + double deletion.
Instead of storing `UniqueID`, `RefPtr<Connection>` should be used, as is business-as-usual with refcounted objects in general.

smfr replied:
I cache an IPC::Connection now.

kkinnunen-apple commented on Source/WebKit/UIProcess/mac/DisplayLink.h:
```diff
@@ -27,70 +27,71 @@
-    // When responsiveness is critical, we send the IPC to a background queue. Otherwise, we send it to the
-    // main thread to avoid unnecessary thread hopping and save power.
-    static void setShouldSendIPCOnBackgroundQueue(bool value) { shouldSendIPCOnBackgroundQueue = value; }
+    void addObserver(Client&, DisplayLinkObserverID, WebCore::FramesPerSecond);
```
add-remove interface coupled with weakptr usage is "belt-and-suspenders".

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4a2e260fd14a1f44ed52f92c94a18cc3e39c2cb1. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/5496

cdumez commented on Source/WebKit/UIProcess/mac/DisplayLink.h:
```diff
@@ -27,70 +27,71 @@
+    };
 
-    void setPreferredFramesPerSecond(IPC::Connection::UniqueID, DisplayLinkObserverID, WebCore::FramesPerSecond);
+    DisplayLink(WebCore::PlatformDisplayID);
```
explicit?

cdumez reviewed (approved):
LGTM

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/dbebc23921182b608d447844b451fc0a8027fe56. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/5496

smfr committed:
Prepare DisplayLink to support UI-side compositing
https://bugs.webkit.org/show_bug.cgi?id=246694
rdar://101296426

Reviewed by Chris Dumez.

The DisplayLink class has clients (currently denoted by IPC::Connection::UniqueID), and each client
has a collection of DisplayLinkObserverID, whose values are only unique in that context of that client.
In addition, each client has some extra state in the form of "wants full speed updates" which is used
for 120Hz scrolling animations and event handling.

With UI-side compositing, we'll need a client in the UI process. So replace IPC::Connection::UniqueID
with a Client class. For web process clients, DisplayLinkProcessProxyClient implements DisplayLink::Client,
and handles the IPC to the web process. WebProcessProxy owns an instance of DisplayLinkProcessProxyClient.
Getting WebProcessProxy's IPC connection isn't thread-safe, so we cache the IPC::Connection on
the main thread. DisplayLink::Client is accessed via a thread-safe CheckedPtr<> to catch lifetime
issues.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
(WebKit::WebProcessPool::startDisplayLink):
(WebKit::WebProcessPool::stopDisplayLink):
(WebKit::WebProcessPool::stopDisplayLinks):
(WebKit::WebProcessPool::setDisplayLinkPreferredFramesPerSecond):
(WebKit::WebProcessPool::setDisplayLinkForDisplayWantsFullSpeedUpdates):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::updateDisplayLinkFrequency):
(WebKit::WebPageProxy::windowScreenDidChange):
* Source/WebKit/UIProcess/WebProcessPool.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::~WebProcessProxy):
(WebKit::WebProcessProxy::connectionWillOpen):
(WebKit::WebProcessProxy::processWillShutDown):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::displayLinkClient):
* Source/WebKit/UIProcess/mac/DisplayLink.cpp:
(WebKit::DisplayLink::addObserver):
(WebKit::DisplayLink::removeObserver):
(WebKit::DisplayLink::removeClient):
(WebKit::DisplayLink::removeInfoForClientIfUnused):
(WebKit::DisplayLink::incrementFullSpeedRequestClientCount):
(WebKit::DisplayLink::decrementFullSpeedRequestClientCount):
(WebKit::DisplayLink::setObserverPreferredFramesPerSecond):
(WebKit::DisplayLink::notifyObserversDisplayWasRefreshed):
(WebKit::DisplayLink::removeObservers): Deleted.
(WebKit::DisplayLink::removeInfoForConnectionIfPossible): Deleted.
(WebKit::DisplayLink::setPreferredFramesPerSecond): Deleted.
* Source/WebKit/UIProcess/mac/DisplayLink.h:
(WebKit::DisplayLink::displayID const):
(WebKit::DisplayLink::setShouldSendIPCOnBackgroundQueue): Deleted.
* Source/WebKit/UIProcess/mac/DisplayLinkProcessProxyClient.cpp: Added.
(WebKit::DisplayLinkProcessProxyClient::setConnection):
(WebKit::DisplayLinkProcessProxyClient::displayLinkFired):
* Source/WebKit/UIProcess/mac/DisplayLinkProcessProxyClient.h: Added.
* Source/WebKit/UIProcess/mac/WebProcessProxyMac.mm:
(WebKit::WebProcessProxy::startDisplayLink):
(WebKit::WebProcessProxy::stopDisplayLink):
(WebKit::WebProcessProxy::setDisplayLinkPreferredFramesPerSecond):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDisplayRefreshMonitor.mm:
(WebKit::RemoteLayerTreeDisplayRefreshMonitor::requestRefreshCallback):

Canonical link: https://commits.webkit.org/255970@main

webkit-commit-queue commented:
Committed 255970@main (0ca300a709ee): <https://commits.webkit.org/255970@main>

Reviewed commits have been landed. Closing PR #5496 and removing active labels.
